sql/schemachanger: fix trigger dependencies on self

spilchen opened the pull request:
## Fix trigger dependencies on self

This change corrects an issue where a trigger on a table that referenced the same table did not properly record its dependency. Previously, such dependencies were missing.

The fix ensures that self-referential trigger dependencies are properly tracked in both the declarative schema changer and the legacy schema changer. 

Fixes #142370.

Release note (bug fix): Fixed an issue where self-referencing triggers did not have their dependencies properly recorded, which could lead to broken dependencies.

## Include trigger name in dependency error message

Previously, when an operation failed due to a dependency, the error message would mention the dependent table but not the specific trigger causing the dependency. This made it harder to understand what exactly was blocking the operation.

This change improves the error message by including the trigger name. This makes the errors more consistent with errors coming from the DSC, which already include the trigger name. For example:

Before:
```
pq: cannot rename relation "sc1.tbl1" because table "tbl1" depends on it
```

After:
```
pq: cannot rename relation "sc1.tbl1" because trigger "tr1" on table "tbl1" depends on it
```

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147018)
<!-- Reviewable:end -->

spilchen committed:
sql/schemachanger: fix trigger dependencies on self

This change corrects an issue where a trigger on a table that referenced
the same table did not properly record its dependency. Previously, such
dependencies were missing.

The fix ensures that self-referential trigger dependencies are properly
tracked in both the declarative schema changer and the legacy schema
changer.

Fixes #142370.

Release note (bug fix): Fixed an issue where self-referencing triggers
did not have their dependencies properly recorded, which could lead to
broken dependencies.

rafiss reviewed (approved):
nice work! the error message change is very useful

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147018#-:-OQnRrcGEvu6zx5BxMFx:b910z58)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao)

___
*[`pkg/sql/rename_table.go` line 309 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147018#-OQnRTas2jEn9xMZbyZK:-OQnRTas2jEn9xMZbyZL:b-ot644b) ([raw file](https://github.com/cockroachdb/cockroach/blob/93ebbc4d8943868a3e5a77fd3e158c3bec6c4f97/pkg/sql/rename_table.go#L309)):*
> ```Go
> 	desc catalog.TableDescriptor,
> 	targetID descpb.ID,
> 	op string,
> ```

super nit: `op` and `typeName` could be `redact.SafeString` here.
___
*[`pkg/sql/rename_table.go` line 322 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147018#-OQnRaZv1bRtcloTJjwL:-OQnRaZv1bRtcloTJjwM:b-e0d3hp) ([raw file](https://github.com/cockroachdb/cockroach/blob/93ebbc4d8943868a3e5a77fd3e158c3bec6c4f97/pkg/sql/rename_table.go#L322)):*
> ```Go
> 					objName,
> 					redact.SafeString(trigger.Name),
> 					redact.SafeString(desc.GetName()),
> ```

nit: in general, we don't usually mark object names as safe for redaction 


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147018)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @rafiss)

___
*[`pkg/sql/rename_table.go` line 309 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147018#-OQnRTas2jEn9xMZbyZK:-OQnjxrI-_IsvIKbMFpa:b-ovwaw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/93ebbc4d8943868a3e5a77fd3e158c3bec6c4f97/pkg/sql/rename_table.go#L309)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: `op` and `typeName` could be `redact.SafeString` here.
</blockquote></details>

Good idea.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: include trigger name in dependency error message

Previously, when an operation failed due to a dependency, the error
message would mention the dependent table but not the specific trigger
causing the dependency. This made it harder to understand what exactly was
blocking the operation.

This change improves the error message by including the trigger name. This
makes the errors more consistent with errors coming from the DSC, which already
include the trigger name. For example:

Before:
  pq: cannot rename relation "sc1.tbl1" because table "tbl1" depends on it

After:
  pq: cannot rename relation "sc1.tbl1" because trigger "tr1" on table "tbl1" depends on it

Epic: none
Release note: none

spilchen commented:
I ran into two issues. 
- DROP DATABASE in the legacy schema was blocked because of the new error I added that prevent drop cascade when triggers are present. I adjusted that so it doesn’t apply. I also added tests to the triggers logic test to drop the database using the legacy schema changer
- RESTORE when a table has a trigger will drop the trigger. But it failed to update the dependencies correctly. This is a larger issue so I opened #147155. I did add some code to avoid this for the common case. But it’s kind of hacky. 

I think we should fix #147155 before this PR is merged. So, I’m going to put this on hold. I don’t like the hacky code that I added.

rafiss reviewed (commented):
great work! it's nice to see the tests coming together.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147018#-:-ORWfaUP7kH9LClLMy3n:b9j0437)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao)

___
*[`pkg/sql/drop_view.go` line 191 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147018#-ORWf09F1b7DDzgx_QkJ:-ORWf09F1b7DDzgx_QkK:b-5xz6s4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0bfd9a12275a9f80302e7dd71448bb14b577ab1/pkg/sql/drop_view.go#L191)):*
> ```Go
> 	}()
> 
> 	// TODO(146722): when dropping the database, we pass false for blockOnTriggerDependency
> ```

nit: this comment says "database," but should it say "view"?
___
*[`pkg/sql/drop_table.go` line 324 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147018#-ORWec4DExkkVb84vqU7:-ORWec4DExkkVb84vqU8:b-pszk9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0bfd9a12275a9f80302e7dd71448bb14b577ab1/pkg/sql/drop_table.go#L324)):*
> ```Go
> 				// the table we are dropping, and the trigger is defined on another table.
> 				// We should only get in here for DROP DATABASE or SCHEMA.
> 				//if !droppingParent {
> ```

nit: was this debugging code?
___
*[`pkg/sql/rename_table.go` line 295 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147018#-ORWf8_X3uYAuv5gYnAj:-ORWf8_X3uYAuv5gYnAk:bp1kkd1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0bfd9a12275a9f80302e7dd71448bb14b577ab1/pkg/sql/rename_table.go#L295)):*
> ```Go
> 	parentID descpb.ID,
> 	desc catalog.TableDescriptor,
> 	op string,
> ```

nit: op and typename can be `redact.SafeString`


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147018)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @rafiss)

___
*[`pkg/sql/drop_table.go` line 324 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147018#-ORWec4DExkkVb84vqU7:-ORXLqFp6Z3aAf3SQhES:b-alq9bm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0bfd9a12275a9f80302e7dd71448bb14b577ab1/pkg/sql/drop_table.go#L324)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: was this debugging code?
</blockquote></details>

Good catch. This was a new assertion that I tried to add, but didn't end up needing it. I'll remove it.
___
*[`pkg/sql/drop_view.go` line 191 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147018#-ORWf09F1b7DDzgx_QkJ:-ORkVVXt9IaP1UbAJvl4:b4qno34) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0bfd9a12275a9f80302e7dd71448bb14b577ab1/pkg/sql/drop_view.go#L191)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: this comment says "database," but should it say "view"?
</blockquote></details>

I did mean database because you can be in here for drop database. We don't actually know the exact context. So, I end up passing false unconditionally. I'll revise the comment to clarify this.
___
*[`pkg/sql/rename_table.go` line 295 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147018#-ORWf8_X3uYAuv5gYnAj:-ORXN4VS-wS0VTNrC8iY:b-z7pvey) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0bfd9a12275a9f80302e7dd71448bb14b577ab1/pkg/sql/rename_table.go#L295)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: op and typename can be `redact.SafeString`
</blockquote></details>

Hmm, I did make this change from an earlier comment you had. It's in the 2nd commit though. I wonder if that was why you didn't see it initially?


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147018#-:-ORqwjA7Am2cpm0AxTQK:b-7nbn20)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao)



<!-- Sent from Reviewable.io -->
